pytensor/tensor/math.py

```diff
@@ -2366,7 +2361,92 @@ def tensordot(a, b, axes=2):
     See the documentation of numpy.tensordot for more examples.
 
     """
-    return _tensordot_as_dot(a, b, axes, dot=dot, batched=False)
+    try:
+        iter(axes)
+    except Exception:
+        axes_a = list(range(-axes, 0))
+        axes_b = list(range(0, axes))
+    else:
+        axes_a, axes_b = axes
+    try:
+        na = len(axes_a)
+        axes_a = list(axes_a)
+    except TypeError:
+        axes_a = [axes_a]
+        na = 1
+    try:
+        nb = len(axes_b)
+        axes_b = list(axes_b)
+    except TypeError:
+        axes_b = [axes_b]
+        nb = 1
+
+    a = as_tensor_variable(a)
+    b = as_tensor_variable(b)
+    as_ = a.shape
+    bra = a.broadcastable
+    ats = a.type.shape
+    nda = a.ndim
+    bs = b.shape
+    brb = b.broadcastable
+    bts = b.type.shape
+    ndb = b.ndim
+    if na != nb:
+        raise ValueError(
+            "The number of axes supplied for tensordot must be equal for each tensor. "
+            f"Got {na} and {nb} respectively."
+        )
+    for k in range(na):
+        ax_a = axes_a[k]
+        ax_b = axes_b[k]
+        if ax_a < 0:
+            axes_a[k] += nda
+        if axes_a[k] < 0 or axes_a[k] >= nda:
+            raise ValueError(
+                f"Supplied axes {ax_a} for first input of tensordot is out of bounds. "
+                f"Input tensor has only ndim={nda}."
+            )
+        if ax_b < 0:
+            axes_b[k] += ndb
+        if axes_b[k] < 0 or axes_b[k] >= ndb:
+            raise ValueError(
+                f"Supplied axes {ax_b} for first input of tensordot is out of bounds. "
+                f"Input tensor has only ndim={ndb}."
+            )
+        if (bra[ax_a] != brb[ax_b]) or (
+            ats[ax_a] is not None and bts[ax_b] is not None and ats[ax_a] != bts[ax_b]
+        ):
+            raise ValueError(
+                "Input arrays have inconsistent broadcastable pattern or type shape along the axes "
+                "must be multiplied and summed with tensordot."
+            )
+        elif ats[ax_a] is None or bts[ax_b] is None:
+            a = Assert(
+                "Input array shape along reduced axes of tensordot are not equal"
+            )(a, eq(a.shape[ax_a], b.shape[ax_b]))
+
+    # Move the axes to sum over to the end of "a"
+    # and to the front of "b"
+    notin = [k for k in range(nda) if k not in axes_a]
+    newaxes_a = notin + axes_a
+    N2 = 1
+    for axis in axes_a:
+        N2 *= as_[axis]
+    newshape_a = (cast(prod([as_[ax] for ax in notin]), "int64"), N2)
```

Comment:
Isn't this the same as (-1, N2)?

Reply:
Maybe yes. I'm not sure if the static shape info gets lost in the process when you put in `(-1, N2)`

Reply:
I imagine it gets as lost as the prod

Reply:
Something we could improve in the Reshape make_node regardless